drivers: usb: Add nRF52840 USBD driver

pawelzadrozniak opened the pull request:
* Add usbd_dc_nrfx shim
* Add nrfx_usbd driver into ext/hal/nordic/nrfx

The shim is based on the previous one usbd_dc_nrf5.
For handling the USBD hardware, tested nrfx_usbd driver from nRF SDK
was used.

Briefly tested examples:
* usb/cdc_acm
* usb/dfu (USB communication only due to flash handling issues)
* usb/hid-mouse
* bluetooth/hci_usb

The driver relies on nrfx (ext/hal/nordic/nrfx) which should be updated/merged in the first place.

Fixes #9628.

pawelzadrozniak committed:
drivers: usb: Add nRF52840 USBD driver

* Add usbd_dc_nrfx shim

The shim is based on the previous one usbd_dc_nrf5.
For handling the USBD hardware, tested nrfx_usbd driver from nRF SDK
was used.

Briefly tested examples:
* usb/cdc_acm
* usb/dfu (USB communication only due to flash handling issues)
* usb/hid-mouse
* bluetooth/hci_usb

Signed-off-by: Paweł Zadrożniak <pawel.zadrozniak@nordicsemi.no>

anangl commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -0,0 +1,1580 @@
+			usbd_evt_put(ev);
+			ep_ctx->read_pending = true;
+			usbd_work_schedule();
+		} else if (p_event->data.eptransfer.status == NRF_USBD_EP_OK) {
```
And what about the other statuses: `NRF_USBD_EP_OVERLOAD` and `NRF_USBD_EP_ABORTED`. Shouldn't these also be handled (at least regarding `ctx->dma_in_use`)?

jfischer-phytec-iot commented:
@pawelzadrozniak please rebase on master (rebased on my own I got two issues: NRF_USBD_ISOSPLIT_Half is NRF_USBD_ISOSPLIT_HALF now and [usb/dc] [ERR] hf_clock_enable: NRF5 HF clock start fail: -68)

carlescufi commented:
Depends on #10148, rebase after that is merged

carlescufi commented:
hey @pawelzadrozniak, #10148 is merged. Can you rebase this?

carlescufi commented:
@pawelzadrozniak there are some checkpatch errors here:
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/23037/5/tests

@finikorg @jfischer-phytec-iot could you take a look at this please?

carlescufi commented:
@rakons would you mind revisiting this PR now that it has been rebased by @pawelzadrozniak ?

anangl commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -0,0 +1,1630 @@
+	 * was not initialized before. This can be ignored.
+	 */
+	(void)nrfx_usbd_disable();
+	(void)nrfx_usbd_uninit();
```
`nrfx_usbd_uninit()` no longer returns a value.

anangl commented on ext/hal/nordic/nrfx_config_nrf52840.h:
```diff
@@ -2977,7 +2977,7 @@
 #ifndef NRFX_USBD_ENABLED
-#define NRFX_USBD_ENABLED 0
+#define NRFX_USBD_ENABLED 1
 #endif
```
```
#ifdef CONFIG_NRFX_USBD
#define NRFX_USBD_ENABLED 1
#endif
```

anangl commented on ext/hal/nordic/nrfx_config_nrf52840.h:
```diff
@@ -2163,7 +2163,7 @@
 #ifndef NRFX_SYSTICK_ENABLED
-#define NRFX_SYSTICK_ENABLED 0
+#define NRFX_SYSTICK_ENABLED 1
 #endif
```
Still not addressed.

carlescufi commented:
@pawelzadrozniak the commit message in the first commit is wrong, since you are no longer adding anything to `ext/`

finikorg commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -0,0 +1,1567 @@
+
+
+
+
```
Those newlines are not needed here and below

finikorg commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -0,0 +1,1567 @@
+	return NULL;
+}
+
+
```
one newline is enough

anangl commented on subsys/usb/usb_device.c:
```diff
@@ -234,7 +234,7 @@ static bool usb_handle_request(struct usb_setup_packet *setup,
 static void usb_data_to_host(void)
 {
-	u32_t chunk = min(MAX_PACKET_SIZE0, usb_dev.data_buf_residue);
+	u32_t chunk = usb_dev.data_buf_residue;
```
This change should be done in a separate commit (currently it is in "drivers: usb: Add nRF52840 USBD driver") and described properly in the commit message.

jfischer-phytec-iot replied:
Maybe it would be better to open own PR for it, I do not sure that every driver can handle it. We should test existing drivers before it is merged.

carlescufi replied:
@jfischer-phytec-iot  @finikorg I think that's probably the safest.
@pawelzadrozniak can you please open a separate PR with this change so we can get that tested and merged separately?

jfischer-phytec-iot commented on subsys/usb/usb_device.c:
```diff
@@ -234,7 +234,7 @@ static bool usb_handle_request(struct usb_setup_packet *setup,
+	u32_t chunk = usb_dev.data_buf_residue;
 
 	/*Always EP0 for control*/
 	usb_dc_ep_write(0x80, usb_dev.data_buf, chunk, &chunk);
```
Is not part of this PR and there is nothing magic in 0x80

jfischer-no reviewed (commented):
I wanted to test it with sample/subsys/usb/testusb, but it fails on set config. I will check with the sniffer later.

```
[194772.826699] usbcore: registered new interface driver usbtest
[194779.379240] usb 9-7.4: new full-speed USB device number 49 using ehci-pci
[194779.490653] usb 9-7.4: New USB device found, idVendor=2fe3, idProduct=0100
[194779.490657] usb 9-7.4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[194779.490660] usb 9-7.4: Product: Zephyr testusb sample
[194779.490663] usb 9-7.4: Manufacturer: ZEPHYR
[194779.490666] usb 9-7.4: SerialNumber: 0.01
[194784.602544] usb 9-7.4: can't set config #1, error -110
```

carlescufi commented:
@rakons @anangl can you re-review without taking into account the extra commit that fixes a USB core issue?

nashif commented:
@pawelzadrozniak please rebase

anangl commented on drivers/clock_control/nrf5_power_clock.c:
```diff
@@ -414,70 +414,32 @@ DEVICE_AND_API_INIT(clock_nrf5_k32src,
-}
-
-static bool usbregstatus_vbusdet_get(void)
+bool nrf5_power_clock_usb_vbusdet(void)
```
I see no point in keeping this function. The USB shim can call the POWER HAL itself. Same applies to `nrf5_power_clock_usb_outrdy`.
`nrf5_power_usb_power_int_enable` is a bit different since it refers to the POWER_CLOCK IRQ, so it should stay here (temporarily, until we find a better solution instead of it),

carlescufi commented:
@anangl could you take one more look at this?

carlescufi reviewed (approved):
@pawelzadrozniak you are changing files in `ext/hal` in the same commit where you add the driver shim. Can you split the changes to `ext/hal` in a separate commit?
Also can you please merge the last commit, the one increasing FIFO_ELEM_COUNT into the driver shim commit?

carlescufi reviewed (changes requested):
Sorry, meant to request changes

anangl commented on drivers/usb/device/Kconfig:
```diff
@@ -53,6 +53,8 @@ config USB_NRF52840
 	select USB_DEVICE_DRIVER
 	select HAS_DTS_USB
+	select NRFX_USBD
+	select NRFX_SYSTICK
```
This should be selected by the NRFX_USBD option itself. Just like NRFX_NFCT selects NRFX_TIMER4.

anangl commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -0,0 +1,1560 @@
+	if (err != NRFX_SUCCESS) {
+		SYS_LOG_ERR(
+			"nRF USBD transfer error (OUT): %d.",
+			err);
```
You could improve the formatting. Now there is no need to break this line into three pieces. Similar story in [line 572](https://github.com/zephyrproject-rtos/zephyr/pull/8787/files#diff-b6600658d4d8775791cddce8653f308cR572).

pawelzadrozniak committed:
ext: hal: nordic: Add kconfig / CMakeLists entries for nrfx_usbd

Adds entries for USBD driver (nrfx_usbd) for kconfig / CMake.
Updates nrf52840 configuration header to enable USBD.

Signed-off-by: Paweł Zadrożniak <pawel.zadrozniak@nordicsemi.no>

carlescufi commented:
@anangl one last pass?
